src/status_im/wallet/choose_recipient/core.cljs

```diff
@@ -72,6 +69,30 @@
  (fn [{:keys [registry ens-name cb]}]
    (ens/get-addr registry ens-name cb)))
 
+(fx/defn resolve-ens-addresses
+  {:events [:wallet.send/resolve-ens-addresses]}
+  [{{:networks/keys [current-network] :wallet/keys [all-tokens] :as db} :db} message origin address]
+  (if address
```

Comment:
this function needs some clarity, as a first step you could put the following into a separate function:
```
{:resolve-address
         {:registry  (get ens/ens-registries (ethereum/chain-id->chain-keyword (get-in constants/default-networks [current-network :config :NetworkId])))
          :ens-name  (get-in message [:function-arguments :address])
          :cb        #(re-frame/dispatch [:wallet.send/resolve-ens-addresses (merge message {:address address}) origin %])}}
```

Reply:
Also not sure how well you know javascript, but I think here it might be good to use promises. It would work like this:
- you make an effect called :resolve-addresses
- `(fn [registry ens-names cb]
         (-> (js/Promise.all (clj->js (mapv #(js/Promise. (fn [resolve reject] (ens/get-addr registry % resolve))) ens-names)
               (.then cb))`
- could also work with reject in case the name is not resolved

The advantage of this would be that the resolutions are done in parallel and it returns as soon as all names are resolved
That also means that you don't need to call this `resolve-ens-addresses` event recursively because you can go straight to `request-uri-parsed` with all you address solved.

wdyt?

Reply:
It makes sense and I have a basic working knowledge of promises on the JavaScript side. I'd thought about that in an earlier iteration of this but just didn't have enough understanding of how to go back and forth between cljs and js to make it work. I'll see if I can rework `resolve-ens-names` per this flow.

Reply:
@yenda I think this is down to my lack of comprehension on how to properly handle calling javascript functions within clojurescript but I'm not able to make something along the lines of what you suggest work.  Below is what I have so far.  Figwheel compiles it without any complaints but I get possible unhandled promise rejection/syntax errors in the ADB logs when trying to run the app with this code in place.  Is it something simple that I'm missing?  Note, this is definitely still complete as it assumes you're getting both 2 addresses in ENS form and doesn't account for parsing path to validate where the ENS addresses are actually found within the URI..  
```(re-frame/reg-fx
 :resolve-addresses
 (fn [registry ens-names message origin]
 (-> js/Promise.all
     (clj->js (mapv (fn [[path ens-name]]
                      js/Promise. (fn [resolve reject]
                                    ens/get-addr registry ens-name #(resolve [path %]))) ens-names)
              .then (fn [[results]]
                      (let [contract-address (get-in results [:address])
                            to-address (get-in results [:function-arguments :address])]
                        (cond-> true
                          true (js/console.log results)
                          contract-address (update-in message [:address] contract-address)
                          to-address (update-in message [:function-arguments :address] to-address)
                          true (re-frame/dispatch [:wallet/request-uri-parsed message origin]))))
              .catch (fn [[error]]
                       (js/console.log error))
              ))))
```

Reply:
hey kudos for keeping up with the good work, don't hesitate to ask questions earlier without spending too much time scratching your head. For starter you have a few missing/missplaced parens:

```clojure
;; this is a threading maco `->`, it means the result of each expression is passed to the next
;; so in this case I want to pass the result of Promise.all to .then and .catch to register my callbacks
(-> (js/Promise.all (clj->js (mapv (fn [[path ens-name]]
                                     (js/Promise.
                                      (fn [resolve reject]
                                        (ens/get-addr registry ens-name #(resolve [path %]))))) ens-names)))
    (.then (fn [[results]]
             (let [contract-address (get-in results [:address])
                   to-address (get-in results [:function-arguments :address])]
;; here you are using another kind of threading macro, a conditional one. What you are doing is
;; saying "here is my expression, at each stage if my condition on the left is true then pass the value of the previous expression as first argument to the expression on the right
               (cond-> true
                 true (js/console.log results) ;;here you are logging `true` and `results` and probably returning nil
                 contract-address (update-in message [:address] contract-address) ;; here you will have an arity error because you are passing the value of the previous expression, nil, as first argument
                 to-address (update-in message [:function-arguments :address] to-address)
                 true (re-frame/dispatch [:wallet/request-uri-parsed message origin])))))
    (.catch (fn [[error]]
              (js/console.log error))))
```
instead of cond you probably want to use a `do`
note that message is an immutable value, `update-in` doesn't change message, it just returns an updated version of message, what you might want to do is return it or do more things with it, but if you just do it in the middle of effectful logging statements, it is not going to have any effect.

Reply:
threading:
```
(-> message ;; my first expression is just the symbol message, provided I have defined it or provided it as an argument to the function that runs this code
      (update-in [:address] contract-address) ;; the macro will insert message here
      ;; if the function only takes one argument I can omit parens
      send ;; imagine send being a side-effect function takes a message as first argument
```

Reply:
@yenda I'm still struggling to make this work.  I've moved parens around and fixed all the bugs I can find but it's still throwing an "undefined is not a function" error when I scan a qr code.
I've traced this code through and verified that it is passing the ens address to the promise and the `get-addr` function within each promise is resolving the address but as far as I can tell, the promises are never actually resolving since the `.then` piece never fires.  Do you see any more misplaced parens?
```
(re-frame/reg-fx
 :resolve-addresses
 (fn [{:keys [registry ens-names message origin]}]
   (-> js/Promise.all
       (clj->js (mapv (fn [[path ens-name]]
                        (js/Promise.
                         (fn [resolve reject]
                           (ens/get-addr registry ens-name #(resolve [path %]))))) ens-names))
       (.then (fn [results]
                (let [contract-address (get-in results [:address])
                      to-address (get-in results [:function-arguments :address])]
                  (-> message
                      contract-address (update-in [:address] contract-address)
                      to-address (update-in  [:function-arguments :address] to-address)
                      (re-frame/dispatch [:wallet/request-uri-parsed origin])))))
       (.catch (fn [error]
                 (js/console.log error))))))
```

Reply:
@yenda Any thoughts on this latest iteration?  I'd love to get this PR finished up but this approach using interop promises is proving quite challenging to make work.  It seems like the `resolve [path %]` is the likely source of the error about calling an undefined function that I'm seeing but I confess I have no idea how to fix it.  It just seems like the get-addr function isn't calling that resolve callback once it gets the ENS name back from the resolver contract.

Reply:
@acolytec3 sorry I missed your previous message, you forgot the parenthesis around (js/Promise.all.. and its argument

Reply:
also could you push your changes I wanted to help you but seems like the last commit doesn't include these changes